pkg/kubelet/container/runtime.go

```diff
@@ -0,0 +1,72 @@
+/*
+Copyright 2015 Google Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package container
+
+import (
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet/volume"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/types"
+)
+
+// Runtime interface defines the interfaces that should be implemented
+// by a container runtime.
+type Runtime interface {
+	// Version returns a map of version information of the container runtime.
+	Version() (map[string]string, error)
+	// ListPods returns a list of Pod. The boolean parameter specifies whether
+	// the runtime returns all pods including those already exited and dead pods(used
+	// for garbage collection).
+	ListPods(all bool) ([]*Pod, error)
+	// RunPod starts all the containers of a pod within a namespace.
+	RunPod(*api.Pod, map[string]volume.Interface) error
+	// KillPod kills all the containers of a pod.
+	KillPod(*api.Pod) error
+	// RunContainerInPod starts a container within the same namespace of a pod.
+	RunContainerInPod(api.Container, *api.Pod, map[string]volume.Interface) error
+	// KillContainerInPod kills a container in the pod.
+	KillContainerInPod(api.Container, *api.Pod) error
+	// GetPodStatus retrieves the status of the pod, including the information of
+	// all containers in the pod.
+	GetPodStatus(*api.Pod) (api.PodStatus, error)
+
+	// TODO(yifan): Pull/Remove images
+}
+
+// Pod is a group of containers, with the status of the pod.
+type Pod struct {
+	// The ID of the pod, which can be used to retrieve a particular pod
+	// from the pod list returned by ListPods().
+	ID types.UID
+	// List of containers that belongs to this pod. It may contains only
+	// running containers, or mixed with dead ones (when ListPods(true)).
+	Containers []*Container
+	// The status of the pod.
+	Status api.PodStatus
```

Comment:
Only thing that worries me with Status in Pod is that for Docker we'll need to inspect every container after we list. This will be a bit expensive. Lets go with this and see how it goes. The abstraction does make it much nicer to use.